File: bokehjs/test/integration/examples/color_scatter.ts
import {display} from "../_util"

import {figure} from "@bokehjs/api/plotting"
import {zip} from "@bokehjs/core/util/array"
import {Random} from "@bokehjs/core/util/random"
import {ndarray} from "@bokehjs/core/util/ndarray"
import {f} from "@bokehjs/api/expr"

describe("Examples", () => {
  it("should support ColorScatter", async () => {
    const random = new Random(1)

    const N = 4000
    const x = f`${random.floats(N)}*${100}`
    const y = f`${random.floats(N)}*${100}`
    const radii = f`${random.floats(N)}*${1.5}`
    const colors = ndarray(zip(f`${50} + ${2}*${x}`, f`${30} + ${2}*${y}`).flatMap(([r, g]) => [r, g, 150]), {shape: [N, 3], dtype: "uint8"})

    const TOOLS = "hover,crosshair,pan,wheel_zoom,zoom_in,zoom_out,box_zoom,undo,redo,reset,tap,save,box_select,poly_select,lasso_select,examine,help"

    const p = figure({tools: TOOLS})
    p.circle({x, y, radius: radii, fill_color: colors, fill_alpha: 0.6, line_color: null})

    await display(p)
  })
})

File: examples/basic/scatters/color_scatter.py
''' A basic scatter. This example plot demonstrates manual colormapping and
many different plot tools.

.. bokeh-example-metadata::
    :apis: bokeh.plotting.figure.circle
    :refs: :ref:`ug_interaction_tools`, :ref:`ug_basic_scatters_markers`
    :keywords: scatter, tools

'''
import numpy as np

from bokeh.plotting import figure, show

N = 4000
x = np.random.random(size=N) * 100
y = np.random.random(size=N) * 100
radii = np.random.random(size=N) * 1.5
colors = np.array([(r, g, 150) for r, g in zip(50+2*x, 30+2*y)], dtype=np.uint8)

TOOLS="crosshair,pan,wheel_zoom,zoom_in,zoom_out,box_zoom,undo,redo,reset,tap,save,box_select,poly_select,lasso_select,examine,fullscreen,help"

p = figure(tools=TOOLS, width=400, height=400)
p.circle(x, y, radius=radii, fill_color=colors, fill_alpha=0.6, line_color=None)

from bokeh.models import Panel, Node
from bokeh.models.dom import ValueOf

hints = Panel(
    elements=[ValueOf(obj=p, attr="hints", formatter="raw", format="html")],
    position=Node(target="frame", symbol="bottom_left"),
    anchor="bottom_left",
    stylesheets=["""
    :host {
        background-color: white;
        border: 1px solid black;
    }
    """,
    ],
)
p.elements.append(hints)

show(p)
